File: packages/auth/src/__tests__/AuthModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { AuthModule } from '../AuthModule.js';

describe('AuthModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let auth: AuthModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    auth = new AuthModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('login', () => {
    it('sends correct module, action and payload', async () => {
      const promise = auth.login({ provider: 'google', scopes: ['email', 'profile'] });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('auth');
      expect(sent.action).toBe('login');
      expect(sent.payload).toEqual({ provider: 'google', scopes: ['email', 'profile'] });

      respondSuccess({ token: 'abc123', refreshToken: null, expiresAt: '2026-12-31', user: null });
      const result = await promise;
      expect(result.token).toBe('abc123');
      expect(result.refreshToken).toBeNull();
    });
  });

  describe('logout', () => {
    it('sends correct module and action', async () => {
      const promise = auth.logout();
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('auth');
      expect(sent.action).toBe('logout');
      respondSuccess();
      await promise;
    });
  });

  describe('getToken', () => {
    it('returns token result', async () => {
      const promise = auth.getToken();
      respondSuccess({ token: 'tok', expiresAt: '2026-12-31' });
      const result = await promise;
      expect(result.token).toBe('tok');
      expect(result.expiresAt).toBe('2026-12-31');
    });
  });

  describe('refreshToken', () => {
    it('returns refreshed login result', async () => {
      const promise = auth.refreshToken();
      respondSuccess({ token: 'new-tok', refreshToken: 'ref', expiresAt: '2027-01-01', user: { id: 'u1', email: 'a@b.com', name: 'A', profileImage: null } });
      const result = await promise;
      expect(result.token).toBe('new-tok');
      expect(result.user?.id).toBe('u1');
    });
  });

  describe('getUser', () => {
    it('returns user result', async () => {
      const promise = auth.getUser();
      respondSuccess({ user: { id: 'u1', email: null, name: 'Test', profileImage: null } });
      const result = await promise;
      expect(result.user?.name).toBe('Test');
    });
  });

  describe('onAuthStateChange', () => {
    it('subscribes to auth state events and returns unsubscribe', () => {
      const received: unknown[] = [];
      const unsub = auth.onAuthStateChange((data) => received.push(data));

      simulateNativeEvent('auth', 'authStateChange', { authenticated: true, user: { id: 'u1', email: null, name: null, profileImage: null } });
      expect(received).toHaveLength(1);
      expect((received[0] as any).authenticated).toBe(true);

      unsub();
      simulateNativeEvent('auth', 'authStateChange', { authenticated: false, user: null });
      expect(received).toHaveLength(1);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = auth.login({ provider: 'google' });
      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Auth failed' },
      };
      transport.simulateIncoming(JSON.stringify(response));
      await expect(promise).rejects.toThrow('Auth failed');
    });
  });
});

File: android/auth/src/test/kotlin/com/devryner/rynbridge/auth/AuthModuleTest.kt
package com.devryner.rynbridge.auth

import com.devryner.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class AuthModuleTest {

    @Test
    fun `login returns login result`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["login"]!!

        val result = handler(mapOf(
            "provider" to BridgeValue.string("google"),
            "scopes" to BridgeValue.array(listOf(BridgeValue.string("email"), BridgeValue.string("profile")))
        ))
        assertEquals("mock-token-123", result["token"]?.stringValue)
        assertEquals("mock-refresh-456", result["refreshToken"]?.stringValue)
        assertEquals("2026-12-31T23:59:59Z", result["expiresAt"]?.stringValue)
        assertNotNull(result["user"]?.dictionaryValue)
    }

    @Test
    fun `login passes provider and scopes`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["login"]!!

        handler(mapOf(
            "provider" to BridgeValue.string("apple"),
            "scopes" to BridgeValue.array(listOf(BridgeValue.string("openid")))
        ))
        assertEquals("apple", provider.lastLoginProvider)
        assertEquals(listOf("openid"), provider.lastLoginScopes)
    }

    @Test
    fun `login without scopes defaults to empty`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["login"]!!

        handler(mapOf("provider" to BridgeValue.string("google")))
        assertEquals(emptyList<String>(), provider.lastLoginScopes)
    }

    @Test
    fun `login missing provider throws`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["login"]!!

        val error = assertThrows(RynBridgeError::class.java) {
            kotlinx.coroutines.test.runTest { handler(emptyMap()) }
        }
        assertEquals(ErrorCode.INVALID_MESSAGE, error.code)
    }

    @Test
    fun `logout returns empty`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["logout"]!!

        val result = handler(emptyMap())
        assertTrue(result.isEmpty())
        assertTrue(provider.logoutCalled)
    }

    @Test
    fun `getToken returns token result`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["getToken"]!!

        val result = handler(emptyMap())
        assertEquals("current-token-789", result["token"]?.stringValue)
        assertEquals("2026-12-31T23:59:59Z", result["expiresAt"]?.stringValue)
    }

    @Test
    fun `refreshToken returns new login result`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["refreshToken"]!!

        val result = handler(emptyMap())
        assertEquals("refreshed-token-000", result["token"]?.stringValue)
        assertEquals("2027-01-01T00:00:00Z", result["expiresAt"]?.stringValue)
    }

    @Test
    fun `getUser returns user data`() = runTest {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        val handler = module.actions["getUser"]!!

        val result = handler(emptyMap())
        val user = result["user"]?.dictionaryValue
        assertNotNull(user)
        assertEquals("user-1", user?.get("id")?.stringValue)
        assertEquals("test@example.com", user?.get("email")?.stringValue)
        assertEquals("Test User", user?.get("name")?.stringValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockAuthProvider()
        val module = AuthModule(provider)
        assertEquals("auth", module.name)
        assertEquals("0.1.0", module.version)
    }

    @Test
    fun `end to end with bridge`() = runTest {
        val transport = MockTransport()
        val bridge = RynBridge(transport = transport, config = BridgeConfig(timeout = 5000L))
        val provider = MockAuthProvider()
        bridge.register(AuthModule(provider))

        val requestJSON = """{"id":"req-1","module":"auth","action":"getToken","payload":{},"version":"0.1.0"}"""
        transport.simulateIncoming(requestJSON)

        transport.awaitSent(1)

        assertEquals(1, transport.sent.size)
        val json = Json { ignoreUnknownKeys = true }
        val response = json.decodeFromString<BridgeResponse>(transport.sent[0])
        assertEquals("req-1", response.id)
        assertEquals(ResponseStatus.success, response.status)
        assertEquals("current-token-789", response.payload["token"]?.stringValue)

        bridge.dispose()
    }
}

private class MockAuthProvider : AuthProvider {
    var lastLoginProvider: String? = null
    var lastLoginScopes: List<String>? = null
    var logoutCalled = false

    override suspend fun login(provider: String, scopes: List<String>): LoginResult {
        lastLoginProvider = provider
        lastLoginScopes = scopes
        return LoginResult(
            token = "mock-token-123",
            refreshToken = "mock-refresh-456",
            expiresAt = "2026-12-31T23:59:59Z",
            user = AuthUser(id = "user-1", email = "test@example.com", name = "Test User", profileImage = null)
        )
    }

    override suspend fun logout() {
        logoutCalled = true
    }

    override suspend fun getToken(): TokenResult {
        return TokenResult(token = "current-token-789", expiresAt = "2026-12-31T23:59:59Z")
    }

    override suspend fun refreshToken(): LoginResult {
        return LoginResult(
            token = "refreshed-token-000",
            refreshToken = null,
            expiresAt = "2027-01-01T00:00:00Z",
            user = null
        )
    }

    override suspend fun getUser(): AuthUser? {
        return AuthUser(id = "user-1", email = "test@example.com", name = "Test User", profileImage = null)
    }
}
